### ISSUE TITLE
Automate test data upload from XCA

### ISSUE THREAD

**Issue by mwinokan** [m2ms/fragalysis-frontend#1588]
@mwinokan please give @tdudgeon some paths to test data. N.B. the original data should be copied in case it is archived to tape

@tdudgeon then please prepare some unit tests based on this

**Comment by mwinokan** [comment]
@tdudgeon here are some `config.yaml` examples that have been used to run XCA recently:


|     Target     |                                            `config.yaml`                                             |
|----------------|------------------------------------------------------------------------------------------------------|
| A71EV2A        | `/dls/labxchem/data/lb32627/lb32627-66/processing/analysis/xchem_align_uploads/config.yaml`          |
| CHIKV_Mac      | `/dls/labxchem/data/lb32633/lb32633-6/processing/analysis/xchemalign/config.yaml`                    |
| FatA           | `/dls/labxchem/data/lb36049/lb36049-1/processing/lb30602-66/processing/analysis/outputs/config.yaml` |
| Flavi_NS5_RdRp | `/dls/labxchem/data/lb32627/lb32627-71/processing/analysis/DENV-ZIKV-xchemalign/config.yaml`         |
| CpKRS          | `/dls/labxchem/data/lb36049/lb36049-7/processing/analysis/xchemalign/config.yaml`                    |

Please let me know if you need any other paths.

Again please also store the original data somewhere else so that we are not affected by unwanted archiving

**Comment by tdudgeon** [comment]
@mwinokan can you confirm that none of those dataset need to be treated as confidential.

**Comment by mwinokan** [comment]
@tdudgeon Some are definitely entirely confidential, and likely more have at least a confidential subset.

**Comment by tdudgeon** [comment]
@mwinokan  I don't have access to the `FatA` or `CpKRS` data.
And BTW there are errors in the `config.yaml` for Flavi_NS5_RdRp (panddas paths are not relative)

**Comment by tdudgeon** [comment]
@mwinokan @ConorFWild There are problems with the A71EV2A data. It looks like a new reference structure (`A71_2A_new_model_x2721_4.1_ref3`) has recently been added to this target, but that PDB file is missing the C-terminal residue compared to the other structures. This seems to cause aligner to blow up. Commenting out that reference doesn't help as that  structures is used as a reference in `assemblies.yaml`.

You get an error like this:
`AssertionError: There are no valid reference positions to align. You may want to check residues numbers are the same between your assembly reference and datasets.`

**Comment by tdudgeon** [comment]
@mwinokan The info for `Flavi_NS5_RdRp` does not seem to be correct. The `config.yaml` file you mention is `/dls/labxchem/data/lb32627/lb32627-71/processing/analysis/DENV-ZIKV-xchemalign/config.yaml` but that file refers to data in `/dls/labxchem/data/lb32633/lb32633-1` which seems to be a completely different target.

**Comment by mwinokan** [comment]
@tdudgeon reports that 2A is currently not working, likely due to Ryan working on an upload_3. @mwinokan will need to ping when the upload_3 is stable

**Comment by mwinokan** [comment]
@tdudgeon asks that @mwinokan points out a couple production targets that are suitable for testing

**Comment by mwinokan** [comment]
@phraenquex reiterates that engineering a process for the data curator to snapshot their data, perhaps it can be part of the uplaoder tool.

**Comment by mwinokan** [comment]
@tdudgeon says that there is a copier tool that can be used to manually snapshot the input data is in place.

@phraenquex says that the copier tool should be part of the XCA workflow (before/after collator) so this is happening behind the scenes.

@tdudgeon says that the copier tool can take some time to run, and may only be need to run once whereas the collator may need to be iterated on a few times.

@tdudgeon and @kaliif can you please spec out a suggested workflow that automatically snapshots the input data, and if the subsequent upload is successful adds it to our test data. Conversely if the collator fails then the snapshot could be cleaned up.

@tdudgeon says that the test data is currently kept locally, @Waztom suggests that this should be on S3 / object store

@phraenquex advises this is higher priority than #1288

**Comment by mwinokan** [comment]
@tdudgeon has not got around to this yet but will investigate this week

**Comment by tdudgeon** [comment]
Here is some initial investigation of the use of the `copier` tool.

Typically it is run like this:

```
python -m xchemalign.copier -c <path-to-config.yaml> -b <path-to-data-eg-/> -m copy -o <path-to-output-dir>
```

This results in all the files needed by collator (except the `manual` files - see below) and copies them to `<path-to-output-dir>` retaining the full relative paths.

Copying for reasonably big target (lb32627-66) takes about 1 minute e.g. that time will be added to the execution time of `collator`.

It looks like this can fairly easily be run programatically from collator, and based on certain assumptions all parameters can be easily inferred from collators `-d <work-diir>` argument. Those assumptions are:
- the `config.yaml` is present and correct
- input files are copied to a directory such as `<work-dir>/inputs`

@mwinokan Can you please confirm how this should behave. I propose:

1. input files copied to `<work-dir>/inputs`
2. a `--no-copy` parameter that says "files have already been copied, just use them"
3. if `--no-copy` is NOT specified then that `<work-dir>/inputs` is deleted and re-copied
4. the `manual` files are NOT copied and must be provided by the user e.g. under `<work-dir>/references` (though that location(s) can be specified in the `config.yaml), just don't put them in `<work-dir>/inputs`  as they will be clobbered (see step 3)

An additional feature/complication could be to version the input files e.g. for `upload_1` a `<work-dir>/inputs_1` dir, for `upload_2` a `<work-dir>/inputs_2` dir.

**Comment by mwinokan** [comment]
@tdudgeon I guess this copying is all on the same shell (not sure how to phrase it, but no SSH transfer?). I recall we decided this should be going to S3. Have you looked into that?

**Comment by tdudgeon** [comment]
Yes, its just direct file copying (there is a SSH mode but that's not relevant here).
I've not really considered what the loader does with the input data, but presumably it's something like when an upload is successful the zipped up inputs and outputs are copied to S3 (though security concerns for private targets needs to be considered).

**Comment by mwinokan** [comment]
@phraenquex points out that the failing tests (due to bad input data) could also be useful as the way that XCA reports failures to the users should be tested.

@alanbchristie points out that there should be fast tests for commit testing, and longer tests that test more edge cases that could be run regularly but less frequently. XCA is lacking a CI system that runs the real tests

@tdudgeon please work with @alanbchristie to get some CI testing working with the test data you have. If they are too large please subset the data, and if you need incremental test datasets please split up the Mpro alignment into separate uploads.

@phraenquex asks how @tdudgeon proposes the automated test data copying will be added to the XCA alignment workflow?

@tdudgeon says that the copying is most effective at the start of the XCA run as the data could change during alignment.

**Bottom line is: @tdudgeon please run the copier tool to S3 after a successful upload & target load**

**Comment by mwinokan** [comment]
@tdudgeon is working on a proof of concept for the copier to pass on to @kaliif

@tdudgeon points out that currently non-Diamond / manual inputs are not copied. @mwinokan says to only take the files that are used in the alignment (see collator logic)

**Comment by mwinokan** [comment]
@tdudgeon has completed the work on the copier:

- Manual data is copied
- Some quality of life improvements for @kaliif to integrate with the upload tool

**Comment by Waztom** [comment]
@kaliif says this almost complete (60%) and estimates that will be complete today/tomorrow.

**Comment by Waztom** [comment]
@tdudgeon asks who else has the Object store key? Tim will ask for a new key, so that we know with 100% confidence who has access.

**Comment by alanbchristie** [comment]
Having automated (CI) access to test date is important. Can we mitigate concerns using cryptography? Individual data files could be written to the object store using a data-specific encryption algorithm and associated key. Or do we just have to work on the assumption that all test files must contain publicly accessible data?

**Comment by kaliif** [comment]
XCA can now upload input data to S3. The default behaviour is to do this automatically on successful upload, but when something goes wrong with the upload, the user can also run the upload with:

```
python -m xchemalign.uploader [-s|--upload-inputs]
```
When `-s/--upload-inputs` is used, all other arguments are ignored, the script skips the fragalysis upload and goes straight to S3 upload.

There's also a parameter `--no-copy` to prevent the copy if necessary (testing, for instance).

Two issues I'd like some opinions on:
1) When the upload is invoked with the `--upload-inputs` option, the script cannot tell if the data has been uploaded to Fragalysis or not. The bucket having a tarball that's not in fragalysis, is it a problem?

2) It's currently configured to get the S3 credentials from env variables, but this means they will be exposed to anyone using the system. Is this an acceptable situation? The user needs to do nothing more than run the `env` command, and they can get access to the secret key.

**Comment by mwinokan** [comment]
@kaliif says that 1. is not likely to be a common issue and @mwinokan agrees that user's will likely not be uploading their inputs unless specifically instructed.

2. is a more serious risk because the dls/software directory is not carefully controlled (students, etc.) and private data uploaded to S3 could be accessed with the credentials. @Waztom agrees this needs to be run by @phraenquex.

@tdudgeon suggests sending the data to S3 via Fragalysis, so that the credentials are not available at DLS. @Waztom agrees this is more sensible

@mwinokan wonders about a write-only key, which would simplify the data transfer and remedy the risk of the credentials being abused

@Waztom says that the DLS data team may have thoughts on an intermediate layer that handles the S3 transfer. @alanbchristie says that a kubernetes pod could be set up where the data can be sent to initially

**@tdudgeon please can you explore the possibility of the write-only keys**

**Comment by mwinokan** [comment]
@tdudgeon has opened a ticket with STFC regarding the write-only keys

@phraenquex says it's worth doing this properly, including telling the Fragalysis backend where to find the data on S3. 

@kaliif please add a path/key to S3 to the TargetExperimentUpload model in the b/e

@tdudgeon suggests having different S3 buckets for each fragalysis stack URL that the data was uploaded to. That means that the URL argument will be required for `--upload-inputs` as well.

@phraenquex additionally says we should request the write-only keys from Fragalysis after authenticating with Keycloak

@phraenquex also wonders whether we should be sending the private data to S3 at all (@Waztom).

@tdudgeon adds that we will need to clearly sign-post that the upload also mirrors input data in S3 in Slack and the documentation

**Comment by mwinokan** [comment]
@tdudgeon has heard back about the write-only keys but only got bounced over to another person at STFC. Still waiting

**Comment by tdudgeon** [comment]
The STFC storage team have responded, saying that AWS S3 ACLs should be fully supported by Echo S3.
See  https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html

**Comment by Waztom** [comment]
@kaliif says this is complete and was pending data policy discussion. 

@Waztom will create an S3 policy ticket.

**Comment by mwinokan** [comment]
@tdudgeon says that this is parked and S3 copying has been disabled until the data policy has been ironed out.

@mwinokan to schedule a discussion when @phraenquex is available to discuss whether write-only keys are sufficient security.

@Waztom and @alanbchristie also raise that anybody with a kubeconfig can access all the data in Fragalysis regardless of how sensitive it is

@Waztom wonders how Diamond legal might cover this, @tdudgeon says as of yet there has not been a CDA with IM.

@tdudgeon says it would be very difficult to restrict the stack admin access

Furthermore, @alanbchristie says that STFC has backdoor access to the storage access

@Waztom says that some CRUK data would be suitable for a properly secured but public Fragalysis instead of a private cluster. @tdudgeon says that this would be a best effort security with legal agreements

**Next step is for @Waztom to confirm whether IM need to be on a separate agreement**

**Comment by mwinokan** [comment]
@phraenquex says, if it's private it shouldn't go to S3

@mwinokan adds that at upload all new projects created in Fragalysis are private by default

There are still a lot of open security questions which need to go via legal. 

@tdudgeon points out a benefit by going to S3 first and then to Fragalysis in that the loading via HTTP is no longer necessary
